mullvad-daemon/src/main.rs

```diff
@@ -916,3 +945,220 @@ fn running_as_admin() -> bool {
+    use windows_service::service_control_handler::{self, ServiceControlHandlerResult,
+                                                   ServiceStatusHandle};
+    use windows_service::service_dispatcher;
+    use windows_service::service_manager::{ServiceManager, ServiceManagerAccess};
+
+    static SERVICE_NAME: &'static str = "MullvadVPN";
+    static SERVICE_DISPLAY_NAME: &'static str = "Mullvad VPN Service";
+
+    pub fn run() -> Result<()> {
+        let windows_directory = ::std::env::var_os("WINDIR").unwrap();
+        let global_temp_directory = PathBuf::from(windows_directory).join("Temp");
+        let main_log_file = Some(global_temp_directory.join("mullvad-daemon-service.log"));
+
+        logging::init_logger(log::LevelFilter::Debug, main_log_file.as_ref(), true)
+            .chain_err(|| "Unable to initialize logger")?;
+        log_version();
+
+        // Start the service dispatcher.
+        // This will block current thread until the service stopped and spawn `service_main` on a
+        // background thread.
+        service_dispatcher::start_dispatcher(SERVICE_NAME, service_main)
+            .chain_err(|| "Failed to start a service dispatcher")
+    }
+
+    define_windows_service!(service_main, handle_service_main);
+
+    pub fn handle_service_main(arguments: Vec<OsString>) {
+        debug!("Started as a system service.");
+
+        let _ = start_service(arguments);
+    }
+
+    fn start_service(_arguments: Vec<OsString>) -> Result<()> {
+        let windows_directory = ::std::env::var_os("WINDIR").unwrap();
+        let tunnel_log_file = PathBuf::from(windows_directory)
+            .join("Temp")
+            .join("mullvad-daemon-openvpn.log");
+
+        let resource_dir = get_resource_dir();
+        let daemon = Daemon::new(Some(tunnel_log_file), resource_dir, true)
+            .chain_err(|| "Unable to initialize daemon")?;
+        let shutdown_handle = daemon.shutdown_handle();
+
+        // Register service event handler
+        let event_handler = move |control_event| -> ServiceControlHandlerResult {
+            match control_event {
+                // Notifies a service to report its current status information to the service
+                // control manager. Always return NO_ERROR even if not implemented.
+                ServiceControl::Interrogate => ServiceControlHandlerResult::NoError,
+
+                ServiceControl::Shutdown | ServiceControl::Stop => {
+                    shutdown_handle.shutdown();
+                    ServiceControlHandlerResult::NoError
+                }
+
+                _ => ServiceControlHandlerResult::NotImplemented,
+            }
+        };
+
+        let status_handle =
+            service_control_handler::register_control_handler(SERVICE_NAME, event_handler)
+                .chain_err(|| "Failed to register a service control handler")?;
+
+        // Chicken or egg problem: `ServiceState::StartPending` is never sent because
```

Comment:
Basically normally we should report `StartPending` and then do initialization. `StartPending` is a state that SCM puts the service into initially. Every time we report a `Pending` state we can tell the system how long it will take to complete the transition to the new state (via `wait_hint`). 

If the daemon initialization can be lengthy then I need to report `StartPending` before creating a `Daemon` and likely to have an extra `channel` so we could register control_handler before the Daemon is created.